provider/gce/environ_network.go

```diff
@@ -1,40 +1,294 @@
+			results = append(results, makeSubnetInfo(
+				iface.ProviderSubnetId,
+				iface.ProviderNetworkId,
+				iface.CIDR,
+				zones,
+			))
+		}
+	}
+	return results, nil
+}
+
+// NetworkInterfaces implements environs.NetworkingEnviron.
+func (e *environ) NetworkInterfaces(instId instance.Id) ([]network.InterfaceInfo, error) {
+	insts, err := e.Instances([]instance.Id{instId})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	envInst, ok := insts[0].(*environInstance)
+	if !ok {
+		// This shouldn't happen.
+		return nil, errors.Errorf("couldn't extract google instance for %q", instId)
+	}
+	googleInst := envInst.base
+	ifaces := googleInst.NetworkInterfaces()
+
+	subnetURLs := make([]string, len(ifaces))
+	for i, iface := range ifaces {
+		subnetURLs[i] = iface.Subnetwork
+	}
+	subnets, err := e.subnetsByURL(subnetURLs...)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	// We know there'll be a subnet for each url requested, otherwise
+	// there would have been an error.
+
+	var results []network.InterfaceInfo
+	for i, iface := range ifaces {
+		subnet, ok := subnets[iface.Subnetwork]
+		if !ok {
+			// Should never happen.
+			return nil, errors.Errorf("no subnet %q found for instance %q", iface.Subnetwork, instId)
+		}
+
+		results = append(results, network.InterfaceInfo{
+			DeviceIndex: i,
+			CIDR:        subnet.CIDR,
+			// The network interface has no id in GCE so it's
+			// identified by the machine's id + its position.
+			ProviderId:        network.Id(fmt.Sprintf("%s/%d", instId, i)),
+			ProviderSubnetId:  subnet.ProviderId,
+			ProviderNetworkId: subnet.ProviderNetworkId,
+			AvailabilityZones: subnet.AvailabilityZones,
+			InterfaceName:     iface.Name,
+			Address:           network.NewScopedAddress(iface.NetworkIP, network.ScopeCloudLocal),
+			InterfaceType:     network.EthernetInterface,
+			Disabled:          false,
+			NoAutoStart:       false,
+			ConfigType:        network.ConfigDHCP,
+		})
+	}
+	return results, nil
+}
+
+func (e *environ) subnetsByURL(urls ...string) (map[string]network.SubnetInfo, error) {
+	// In GCE all the subnets are in all AZs.
+	zones, err := e.zoneNames()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	networks, err := e.networkIdsByURL()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	// We don't want to get all subnets if no urls are passed.
+	urlSet := includeSet{items: set.NewStrings(urls...)}
+	allSubnets, err := e.gce.Subnetworks(e.cloud.Region)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	results := make(map[string]network.SubnetInfo)
+	for _, subnet := range allSubnets {
+		networkId, ok := networks[subnet.Network]
+		if !ok {
+			return nil, errors.NotFoundf("network %q for subnet %q", subnet.Network, subnet.Name)
+		}
+		if urlSet.Include(subnet.SelfLink) {
+			results[subnet.SelfLink] = makeSubnetInfo(
+				network.Id(subnet.Name),
+				network.Id(networkId),
+				subnet.IpCidrRange,
+				zones,
+			)
+		}
```

Comment:
Do you mean a sanity check to make sure GCE aren't breaking their own rules? I could see that being useful if an interface was marked with the wrong CIDR because two subnets had the same URL. But I'm not sure it's worth restructuring the code to check for. What about a sanity check that the ip address is contained in the CIDR higher up?